enhancements/ocp-coreos-layering.md

```diff
@@ -0,0 +1,479 @@
+---
+title: OpenShift CoreOS Layering
+authors:
+  - "@yuqi-zhang"
+  - "@jkyros"
+  - "@mkenigs"
+  - "@kikisdeliveryservice"
+  - "@cheesesashimi"
+  - "@cgwalters"
+  - "@darkmuggle (emeritus)"
+reviewers:
+  - "@mrunalp"
+approvers:
+  - "@sinnykumari"
+  - "@mrunalp"
+creation-date: 2021-10-19
+last-updated: 2022-02-09
+tracking-link:
+  - https://issues.redhat.com/browse/GRPA-4059
+---
+
+# OpenShift Layered CoreOS (PROVISIONAL)
+
+**NOTE: Nothing in this proposal should be viewed as final.  It is highly likely that details will change.  It is quite possible that larger architectural changes will be made as well.**
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Change RHEL CoreOS as shipped in OpenShift to be a "base image" that can be used as in layered container builds and then booted.  This will allow custom 3rd party agents delivered via RPMs installed in a container build.  The MCO will roll out and monitor these custom builds the same way it does for the "pristine" CoreOS image today.
```

Comment:
Currently, AFAIK the licensing of CoreOS does not allow reusing it for other purposes (FIXME). Will this change with this proposal?

Reply:
There is (to the best of my knowledge) no *licensing* restrictions - "RHEL CoreOS" is just not a separate thing from OpenShift.

I think this question is really addressed by https://github.com/openshift/enhancements/pull/1032/files#diff-d56db1c5d442467039e4ac4e6d118edd49185320a947906e8fd6973cbbd5f1abR60

But, the underlying non-OCP functionality of "boot a container" is in https://fedoraproject.org/wiki/Changes/OstreeNativeContainer and https://github.com/coreos/enhancements/blob/main/os/coreos-layering.md and will ship in the `rpm-ostree` in RHEL.  Truly productizing that is a whole separate thing.  I pushed https://github.com/openshift/enhancements/pull/1032/commits/d25534f6ecb1a27d76b7f542ac871fe31599cf75 related to that.